### PR TITLE
Enable tests on travis; libsolid-rs: test login automatically, introduce client trait

### LINKED ISSUE #8: libsolid-rs: shouldn't rely on interactive tests
We should find alternatives for all interactive tests, tracking the progress here:

- [ ] interactive login to solid.community

### LINKED ISSUE #14: Enable CI tests
With #8 out of the way we can start running the test(s) on CI.

We need to make some solid.community test account credentials available on for the CI job (see #13).

### PR DESCRIPTION
**Pull request by steveej**
Introduce initial support for automatically testing the login flow in
the client. This requires restructurung the login flow and introduce
custom requests on behalf of the user.

Moving some of the generic code from the `auth` into the root library
code, preparing for separation of logic for implementing the various
client features, as well as preparing for the deserializing and
serializing of client state.

Resolves #8.
Resolves #14.

### CODE REVIEW

File: libsolid-rs/src/auth.rs [hunk 1]
```diff
@@ -382,20 +368,312 @@ pub mod webid_oidc {
...
         use super::*;
 
         #[test]
         #[cfg(feature = "test-net")]
-        #[cfg(feature = "test-interactive")]
-        fn interactive_login_to_solid_community() -> Result<()> {
-            let callback_bind_addr = String::from("127.0.0.1:36666");
-
+        fn login_to_solid_community() -> Result<()> {
             let provider_url = "https://solid.community".to_string();
```
Comment (steveej) [inline libsolid-rs/src/auth.rs]: Should we use an env var for this as well? That way the test environment has free choice which server to test against.

Comment (mohammed-ali-1) [reply libsolid-rs/src/auth.rs]: Yes, definitely! It'd be nice if the test can pickup the `provider_url` from the environment. This way we can easily also test against `localhost` or other providers.

File: libsolid-rs/src/auth.rs [hunk 2]
```diff
@@ -59,33 +85,43 @@ pub mod webid_oidc {
 
     /// Client for talking to one WebIdOIDC provider and can handles multiple users' access credentials.
```
Comment (mohammed-ali-1) [inline libsolid-rs/src/auth.rs]: typo with **handles**. Should be handle.

Comment (steveej) [inline libsolid-rs/src/auth.rs]:
```suggestion
     /// Client for talking to one WebIdOIDC provider on behalf of potentially multiple users.
```

Comment (mohammed-ali-1) [reply libsolid-rs/src/auth.rs]: Thanks :)

**Comment by mohammed-ali-1** [comment]
@steveeJ Can you please explain what this function does?

https://github.com/solidrive/solidrive/blob/8533b16750cf95bb4e81db584426286437db49df/libsolid-rs/src/auth.rs#L446

I don't feel like I adequately understand it.

File: libsolid-rs/src/auth.rs (hunk 1 above)
Comment (steveej) [reply libsolid-rs/src/auth.rs]: it does now

**Comment by steveej** [comment]
> Can you please explain what this function does?
> 
> https://github.com/solidrive/solidrive/blob/8533b16750cf95bb4e81db584426286437db49df/libsolid-rs/src/auth.rs#L446

The function you linked is used as the request handler for incoming requests on the callback server. It parses the query parameters into a map and sends it through a channel, on which the other thread of the login procedure is waiting. The query parameters contain a code and a nonce which are part of the OIDC flow.

**Comment by mohammed-ali-1** [comment]
> The function you linked is used as the request handler for incoming requests on the callback server. It parses the query parameters into a map and sends it through a channel, on which the other thread of the login procedure is waiting. The query parameters contain a code and a nonce which are part of the OIDC flow.

Ok! I think I understand it better now. Thanks for the explanation 👍🏻.
My Rust is not quite there yet, still figuring it out 😄

**Review by mohammed-ali-1 — APPROVED** [review]

File: libsolid-rs/src/auth.rs (hunk 1 above)
Comment (mohammed-ali-1) [reply libsolid-rs/src/auth.rs]: AWESOME! 🎉

**Pull request merged**